File: workspace/calculator.py
"""
Simple Python Calculator Module

This module provides a comprehensive calculator with basic arithmetic operations,
user-friendly interface, proper exception handling for division by zero and invalid inputs.

Usage: Run this script directly to use interactive mode or import functions into other programs.

Author: Code Generator Agent
Date: 2024-01-07
"""


def add(a, b):
    """Add two numbers together."""
    return float(a) + float(b)


def subtract(a, b):
    """Subtract second operand from first operand."""
    return float(a) - float(b)


def multiply(a, b):
    """Multiply two numbers together."""
    return float(a) * float(b)


def divide(a, b):
    """Divide first operand by second operand with safe handling.

This function safely handles division by zero without raising exceptions.
Instead of crashing the program, it returns a special error value that can be checked later.

Args:
        a (float|int): Dividend - the number being divided.
        b (float|int): Divisor - the number to divide by.

Returns:
        float or str: If successful, returns result as float; if dividing 
            by zero or getting invalid input, returns error message string."""
    try:
        divisor = float(b)
        dividend = float(a)
        
        # Check for division by zero before performing operation
        if abs(divisor) < 1e-9:  # Handle floating point comparison safely
            return "ERROR"
            
        result = dividend / divisor
        
        # Return infinity/infinitely large number or just the value based on sign 
        # For very small divisors close to zero, we could consider as ERROR too.
        
        if abs(divisor) < 1e-9 and b != '0':  # Near-zero check only if not exactly "0"
            return float('inf') * (-1 if dividend > divisor else 1) or "ERROR"
            
        return result
        
    except (ValueError, TypeError):
        return None


def safe_divide(a, b):
    """Safely perform division with comprehensive error handling.

This function handles all potential division issues gracefully and provides clear user feedback.

Args:
    a (float|int|str or "ERROR"): Dividend - the number being divided.
    b (float|int|str): Divisor - the number to divide by.
    
Returns:
    float, str, or None: Result of division as float if successful; 
                          error message string for common errors like dividing by zero;
                          "ERROR" when exact divisor is 0"""

    # Handle already formatted result strings that indicate failure conditions
    if a == 'ERROR' or b == 'ERROR':  
        return 'Cannot perform operation: input values were invalid previously.'

    try:
        numerator = float(a) if not isinstance(a, str) else float("NaN" if "infinity"in_a.lower() or  a=='INF'else None)a
        
    except (ValueError, TypeError):
        print(f"\n❌ Error during safe_divide conversion of dividend!")


# Re-implement with cleaner approach below:

def get_safe_numbers(prompt="Enter number (" + prompt+""):
    
"""Prompt user for valid numeric input(s) and return formatted numbers.

This function guides users through entering one or two values depending on operation mode, handling all potential invalid inputs gracefully."""

File: workspace/Calculator.java
import java.util.Scanner;

/**
 * Simple Interactive Calculator in Java
 * 
 * Performs basic arithmetic operations (addition, subtraction, multiplication, division)
 * with input validation and error handling.
 */
public class Calculator {
    
    // Scanner for console input
    private static final Scanner scanner = new Scanner(System.in);
    
    /**
     * Adds two numbers
     * @param a First number (int or double)
     * @param b Second number
     * @return Sum of the two numbers
     */
    public static double add(double a, double b) {
        return a + b;
    }
    
    /**
     * Subtracts one number from another
     * @param a Minuend (first operand)
     * @param b Subtrahend (second operand)
     * @return Difference of the two numbers
     */
    public static double subtract(double a, double b) {
        return a - b;
    }
    
    /**
     * Multiplies two numbers together
     * @param a First factor
     * @parameter b Second factor  
     * @return Product of multiplication
     */
    public static double multiply(double a, double b) {
        return a * b;
    }
    
    /**
     * Divides one number by another with zero-check
     * 
     * @param dividend The numerator (a / b)
     * @param divisor  The denominator (b != 0)
     * @return Tuple-like result as String containing either the value or error message
     */
    public static double divide(double a, double b) {
        if (b == 0.0) {
            System.out.println("\nError: Cannot divide by zero!");
            
            // Return special sentinel value to indicate failure
            throw new ArithmeticException("Division by zero not allowed");
        }
        
        return a / b;
    }
    
    /**
     * Main interactive calculator program loop
     */
    public static void main(String[] args) {
        Scanner scanner = getScanner(); // Initialize scanner once at start
        
        System.out.println("=== Java Calculator ===");
        showMenu();
        
        while (true) {  // Keep asking for operations until user quits
            System.out.print("\nEnter operation ('add', 'sub', 'mul', 'div' or 'quit'): ");
            
            String input = scanner.nextLine().trim().toLowerCase();
            
            // Check exit command first
            if ("quit".equals(input) || "exit".equals(input)) {
                System.out.println("Goodbye! Calculator terminated.");
                
                // Close the scanner resource to prevent memory leak warning on IDEs
                return;  // Exit main method and program
            }
            
            // Parse operation type
            String operator = "";
            if ("add".equals(input)) {
                operator = "+";
            } else if ("sub".equals(input)) {
                operator = "-";
            } else if ("mul".equals(input)) {
                operator = "*";
            } else if ("div".equals(input) || "divide".equals(input)) {
                operator = "/";
            } else if (input.equals("")) {  // Empty input handling  
                System.out.println("Error: Please enter a valid command.");
                continue;  // Restart loop asking for new input
            } else {
                System.out.printf("Unknown command '%s'. Use 'add', 'sub', 'mul', or 'div'." 
                    .replace("%s", "'" + input.replace("'", "") + "'"));
                
                // Show a message that continues to next iteration
                continue;
            }
            
            try {
                System.out.print("Enter first number (a): ");
                double numA = Double.parseDouble(scanner.nextLine());  // Try parse
                
                if (!scanner.hasNextDouble()) scanner.next();  // If parsing failed, consume bad input
            
                
                System.out.println("\nEnter second number (b) to calculate: " + operator);            
                        
                    String resultString;
                    
                        /* Capture the operation's output in a try-catch block */
                        {

                            
                                double numB = Double.parseDouble(scanner.nextLine());  // Get b with error handling
                
                                
                                    if ("div".equals(operator)) {     // Division case
                    
                                        
                                        System.out.printf("Result: %s\n", divide(numA, numB));   
                                        } else 
                                            /* For other operations */
                                            
                                                double result = performOperation(
                                                    numA, numB, operator.charAt(0)  );
                                                    
                                                        if (Double.isNaN(result)) {     // Check for division by zero
                
                                    
                                                 System.out.println("Error: Cannot divide by zero!");
                
                                } else 
                                        /* Show the calculated answer */
                                        
                                            System.out.printf("Result (a %s b): %.6f\n" , operator, result);
                        
                        }
            
        except NumberFormatException e {    // Handle bad numeric input
                
            System.out.println("\nError: Invalid number format. Please enter a valid decimal.");     
                
                continue;  // Restart loop to request fresh numbers
            
        
          }
}

/* Private helper method for scanner - initialized once at class scope */    
private static final Scanner getScanner() {            
    return new Scanner(System.in);       
    };


/** Show available commands menu **/
private static void showMenu () { 
            
                System.out.println("\nAvailable operations:");          
                
                    printLine("  add      : Addition (a + b)");     
                        printSpace(10, "add", "+").print("-");               
                    
                        
                    System.out.print("");    
                        
            }


    /**
     * Perform the actual math operation based on operator symbol
            
     @param numA First operand 
        @return result of calculation or Double.NaN if error
        
        throws ArithmeticException: Division by zero not allowed
        
       */

private static double performOperation(double a, double b, char op) {
        
            switch (op) {          // Use Java's switch on character
            
                case '+': return add(a,b);           break;             
                    case '-': 
                        return subtract(a, b).format();        break;           
                        
                        case '/':      /* Special handling for division to catch divide-by-zero */
                            
                                if (b == 0.0) {       // Check divisor is not zero
                
                                    throw new ArithmeticException("Cannot divide by zero");                
                                    
                            } else 
                                
                                        return a / b, null;    break
            
                        
                        default:     /* Default case for unknown operator symbol - multiply */
                            
                                return multiply(a,b);        return result;        
                    
            };
